src/eca/features/chat.clj

```diff
@@ -42,12 +42,238 @@
     (db/update-workspaces-cache! @db*)))
 
 (defn ^:private assert-chat-not-stopped! [{:keys [chat-id db*] :as chat-ctx}]
+  ;; TODO: Should :stoping be :stopping?
   (when (identical? :stoping (get-in @db* [:chats chat-id :status]))
     (finish-chat-prompt! :idle chat-ctx)
     (logger/info logger-tag "Chat prompt stopped:" chat-id)
     (throw (ex-info "Chat prompt stopped" {:silent? true
                                            :chat-id chat-id}))))
 
+;;; Helper functions for tool call state management
+
+(defn ^:private get-tool-call-state
+  "Get the complete state map for a specific tool call."
+  [db chat-id tool-call-id]
+  (get-in db [:chats chat-id :tool-calls tool-call-id]))
+
+(defn ^:private get-active-tool-calls
+  "Returns a map of tool calls that are still active.
+
+  Active tool calls are those not in the following states: :completed, :rejected, :stopped."
+  [db chat-id]
+  (->> (get-in db [:chats chat-id :tool-calls] {})
+       (remove (fn [[_ state]]
+                 (#{:completed :rejected :stopped} (:status state))))
+       (into {})))
+
+;;; Event-driven state machine for tool calls
+
+(def ^:private tool-call-state-machine
```

Comment:
Would be nice to have good unit tests for this state machine

Reply:
Unit tests completed.